plutus-core/plutus-ir/src/PlutusIR/Transform/Rewrite.hs

```diff
@@ -0,0 +1,36 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE PatternSynonyms #-}
+module PlutusIR.Transform.Rewrite where
+
+import PlutusCore.Default
+import PlutusIR.Analysis.Builtins
+import PlutusIR
+import Data.Monoid
+
+import Control.Lens
+
+userRewrite :: (Semigroup a, t ~ Term tyname name uni fun a)
+            => BuiltinsInfo uni fun
+            -> t
+            -> t
+userRewrite bi t =
+    let RewriteRules f = bi^.rewriteRules
+    in transformOf termSubterms (appEndo $ getDual f) t
+
+defaultUniRewriteRules :: RewriteRules DefaultUni DefaultFun
+defaultUniRewriteRules = RewriteRules $ combineRules
+    -- rules are applied from left to right because of Dual
+    [ decodeEncode
+    ]
+  where
+    combineRules = foldMap (Dual . Endo)
```

Comment:
If it were me, I'd probably make `RewriteRules` instead hold a `[Term -> Term]`, and then apply them here. It doesn't matter too much though, I guess.